### PR TITLE
Adding multiplication arithmetic for `RPN` function framework

### PR DESCRIPTION
**Pull request by iosmanthus**
~~TODO: add `multiply_mapper`~~
Apply `proc_macro` to generate `plus/minus/...mapper` 

* `RealMultiply`
* `IntMultiply`
* `IntUintMultiply`

Signed-off-by: Iosmanthus Teng <myosmanthustree@gmail.com>

## What have you changed? (mandatory)
Adding multiplication arithmetic

## What are the type of changes? (mandatory)
- New feature (a change which adds functionality)

## How has this PR been tested? (mandatory)
passed unit test

### CODE REVIEW

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 1]
```diff
@@ -895,4 +938,139 @@ mod tests {
...
+                    .push_param(rhs)
+                    .evaluate(ScalarFuncSig::MultiplyReal)
+                    .unwrap()
+            );
+        }
+
+        let should_fail = unsafe {
+            vec![
+                (
+                    Real::unchecked_new(std::f64::MAX),
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: Please follow the rest of code by using `Real::new(..).ok()`

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 2]
```diff
@@ -895,4 +938,139 @@ mod tests {
...
+    #[test]
+    fn test_int_multiply() {
+        let should_pass = vec![(Some(187), 11, 17), (Some(3), -1, -3)];
+        for (expected, lhs, rhs) in should_pass {
+            assert_eq!(
+                expected,
+                RpnFnScalarEvaluator::new()
+                    .push_param_with_field_type(
+                        lhs,
+                        FieldTypeBuilder::new().tp(FieldTypeTp::LongLong)
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: I guess you can write `FieldTypeTp::LongLong` directly.

File: src/coprocessor/dag/rpn_expr/mod.rs [hunk 3]
```diff
@@ -207,8 +210,8 @@ fn map_pb_sig_to_rpn_func(value: ScalarFuncSig, children: &[Expr]) -> Result<Rpn
         ScalarFuncSig::InDuration => compare_in_fn_meta::<Duration>(),
         ScalarFuncSig::InJson => compare_in_fn_meta::<Json>(),
         _ => return Err(box_err!(
-            "ScalarFunction {:?} is not supported in batch mode",
-            value
+                "ScalarFunction {:?} is not supported in batch mode",
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/mod.rs]: Revert this change.

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 4]
```diff
@@ -314,6 +314,49 @@ impl ArithmeticOp for DecimalMultiply {
...
+pub struct IntMultiply;
+
+impl ArithmeticOp for IntMultiply {
+    type T = Int;
+    fn calc(lhs: &Int, rhs: &Int) -> Result<Option<Int>> {
+        lhs.checked_mul(*rhs)
+            .ok_or_else(|| Error::overflow("INT", &format!("({} * {})", lhs, rhs)).into())
+            .map(Some)
+    }
+}
```
Comment (mapleFU) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: In original `multiply_int` it decide what to do by checking the `signed`/`unsigned` type of `lhs` and `rhs`.
So should we consider using `map_int_sig(value, children, multiply_mapper)?` and implement different type of muliple functions here? @breeswish

Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: Yes. There is a TODO in the PR description: TODO: add multiply_mapper

File: src/coprocessor/dag/rpn_expr/mod.rs (hunk 3 above)
Comment (iosmanthus) [reply src/coprocessor/dag/rpn_expr/mod.rs]: my `vim` formatted my code despite the `#[rustfmt::skip]` :laughing:

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 5]
```diff
@@ -329,9 +397,9 @@ impl ArithmeticOp for IntDivideInt {
 }
 
 #[derive(Debug)]
-pub struct IntDivideUint;
+pub struct IntUintDivide;
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: This change does not relate to this PR.

File: src/coprocessor/dag/rpn_expr/mod.rs [hunk 6]
```diff
@@ -57,41 +58,9 @@ fn compare_mapper<F: CmpOp>(lhs_is_unsigned: bool, rhs_is_unsigned: bool) -> Rpn
...
-    match (lhs_is_unsigned, rhs_is_unsigned) {
-        (false, false) => arithmetic_fn_meta::<IntDivideInt>(),
-        (false, true) => arithmetic_fn_meta::<IntDivideUint>(),
-        (true, false) => arithmetic_fn_meta::<UintDivideInt>(),
-        (true, true) => arithmetic_fn_meta::<UintDivideUint>(),
-    }
-}
+// Here it will generate plus/minus/multiply/divide/mod _mapper()
+// which will generate corresponding RpnFnMeta according to children's unsigned flagss
+int_sig_mapper! { Plus, Minus, Multiply, Divide, Mod }
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/mod.rs]: What about compare_mapper?

File: src/coprocessor/dag/rpn_expr/mod.rs (hunk 3 above)
Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/mod.rs]: You can undo the indent.

File: components/cop_codegen/src/lib.rs [hunk 7]
```diff
@@ -35,3 +36,40 @@ pub fn rpn_fn(attr: TokenStream, input: TokenStream) -> TokenStream {
         Err(e) => TokenStream::from(e.to_compile_error()),
     }
 }
+
+#[proc_macro]
+pub fn int_sig_mapper(input: TokenStream) -> TokenStream {
```
Comment (breeswish) [inline components/cop_codegen/src/lib.rs]: Prefer a trait / generic way instead of procedural macro way.

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 8]
```diff
@@ -895,4 +963,117 @@ mod tests {
...
+
+    #[test]
+    fn test_real_multiply() {
+        let should_pass = vec![(
+            Real::new(-0.0101001).ok(),
+            Real::new(1.01001).unwrap(),
+            Real::new(-0.01).unwrap(),
+        )];
+
+        for (expected, lhs, rhs) in should_pass {
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: please keep the same order as the rest of the test: `(lhs, rhs, expected)`

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 9]
```diff
@@ -895,4 +963,117 @@ mod tests {
...
+                "{} * {} should fail",
+                lhs,
+                rhs
+            );
+        }
+    }
+
+    #[test]
+    fn test_int_multiply() {
+        let should_pass = vec![(Some(187), 11, 17), (Some(3), -1, -3)];
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: Add a MIN * 1 case which is covered previously.

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 10]
```diff
@@ -895,4 +963,117 @@ mod tests {
...
+                "{} * {} should fail",
+                lhs,
+                rhs
+            );
+        }
+    }
+
+    #[test]
+    fn test_int_uint_multiply() {
+        let should_pass = vec![(Some(std::i64::MAX), std::i64::MAX, 1), (Some(21), 3, 7)];
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: How about `negative signed int * uint` ? This case is missing in previous implementation but is also missing in the new impl.

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 11]
```diff
@@ -895,4 +963,117 @@ mod tests {
...
+                            .flag(FieldTypeFlag::UNSIGNED)
+                    )
+                    .evaluate::<Int>(ScalarFuncSig::MultiplyIntUnsigned)
+                    .is_err(),
+                "{} * {} should fail",
+                lhs,
+                rhs
+            );
+        }
+    }
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: missing uint * uint case?

File: src/coprocessor/dag/rpn_expr/mod.rs [hunk 12]
```diff
@@ -170,6 +179,9 @@ fn map_pb_sig_to_rpn_func(value: ScalarFuncSig, children: &[Expr]) -> Result<Rpn
         ScalarFuncSig::MinusReal => arithmetic_fn_meta::<RealMinus>(),
         ScalarFuncSig::MinusDecimal => arithmetic_fn_meta::<DecimalMinus>(),
         ScalarFuncSig::MultiplyDecimal => arithmetic_fn_meta::<DecimalMultiply>(),
+        ScalarFuncSig::MultiplyInt => map_int_sig(value, children, multiply_mapper)?,
+        ScalarFuncSig::MultiplyIntUnsigned => arithmetic_fn_meta::<IntUintMultiply>(),
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/mod.rs]: It should be UintUintMultiply?

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 13]
```diff
@@ -895,4 +963,175 @@ mod tests {
             assert!(output.is_err(), "lhs={:?}, rhs={:?}", lhs, rhs);
         }
     }
+
+    #[test]
+    fn test_real_multiply() {
+        let should_pass = vec![(
+            Real::new(1.01001).unwrap(),
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: Maybe 1.01001f64 is enough

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 14]
```diff
@@ -314,6 +314,74 @@ impl ArithmeticOp for DecimalMultiply {
...
+}
+
+#[derive(Debug)]
+pub struct IntIntMultiply;
+
+impl ArithmeticOp for IntIntMultiply {
+    type T = Int;
+    fn calc(lhs: &Int, rhs: &Int) -> Result<Option<Int>> {
+        lhs.checked_mul(*rhs)
+            .ok_or_else(|| Error::overflow("INT", &format!("({} * {})", lhs, rhs)).into())
```
Comment (mapleFU) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: INT in MySQL is 4 bytes, and BIGINT is 8 bytes. So I think "BIGINT" may be more proper.
